devdoc/channel.md

```diff
@@ -0,0 +1,274 @@
+
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+```
+
+### channel_create
+```c
+    MOCKABLE_FUNCTION(, CHANNEL_HANDLE, channel_create)
+```
+
+`channel_create` creates the channel and returns it.
+
+**SRS_CHANNEL_43_001: [** `channel_create` shall call `THANDLE_MALLOC` with `dispose` as `channel_dispose`. **]**
+
+**SRS_CHANNEL_43_056: [** `channel_create` shall set the `state` to `CHANNEL_READY` and `EPOCH` to `0` by calling `interlocked_exchange`. **]**
+
+**SRS_CHANNEL_43_002: [** If there are any failures, `channel_create` shall fail and return `NULL`. **]**
+
+### channel_dispose
+```c
+    void channel_dispose(CHANNEL* channel)
+```
+
+`channel_dispose` disposes the given `channel`.
+
+**SRS_CHANNEL_43_004: [** `channel_dispose` shall obtain the current `state` of the channel by calling `interlocked_add`. **]**
+
+**SRS_CHANNEL_43_058: [** If the current `state` is `CHANNEL_PULLED`, `channel_dispose` shall call the `pull_callback` given to `channel_pull` with `pull_context` as the `pull_context` given to `channel_pull`, `data` as `NULL`, and `result` as `CHANNEL_CALLBACK_RESULT_ABANDONED`. **]**
+
+**SRS_CHANNEL_43_006: [** If the current `state` is `CHANNEL_PUSHED`, `channel_dispose` shall: **]**
+
+  - **SRS_CHANNEL_43_044: [** call the `push_callback` given to `channel_push` with `push_context` as the `push_context` given to `channel_push`, and `result` as `CHANNEL_RESULT_ABANDONED`. **]**
+
+  - **SRS_CHANNEL_43_045: [** release the reference to `data` stored in `channel` by calling `THANDLE_ASSIGN`. **]**
+
+### channel_pull
+```c
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_pull, THANDLE(CHANNEL), channel, PULL_CALLBACK, pull_callback, void*, pull_context, THANDLE(ASYNC_OP)*, out_op_pull);
+```
+
+`channel_pull` registers the given `pull_callback` to be called when there is data to be consumed.
+
+**SRS_CHANNEL_43_007: [** If `channel` is `NULL`, `channel_pull` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_008: [** If `pull_callback` is `NULL`, `channel_pull` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_009: [** If `out_op_pull` is `NULL`, `channel_pull` shall fail and return `CHANNEL_RESULT_INVALID_ARGS`. **]**
+
+**SRS_CHANNEL_43_010: [** `channel_pull` shall call `interlocked_add` to obtain the `state` of the `channel`. **]**
+
+**SRS_CHANNEL_43_011: [** If the obtained `state` is `CHANNEL_PULLING` or `CHANNEL_PULLED`, `channel_pull` shall fail and return `CHANNEL_RESULT_REFUSED`. **]**
+
+**SRS_CHANNEL_43_059: [** If the obtained `state` is `CHANNEL_PUSHING`, `channel_pull` shall wait for the `state` to change by calling `InterlockedHL_WaitForNotValue`. **]**
+
+**SRS_CHANNEL_43_012: [** If the obtained `state` is `CHANNEL_PUSHED`, `channel_pull` shall : **]**
+
+  - **SRS_CHANNEL_43_060: [** store `push_callback`, `push_context` and `data` given to `channel_push`. **]**
```

Comment:
>store  [](http://example.com/codeflow?start=30&length=6)

Not clean, we should say that we're taking a snapshot on the stack of these values